# Study of NaNs

In this notebook I focused on some features from the 'weather data' with missing values, and I try to find the best way to fill up those NaNs values.

I studied 3 different groups of features here : all the features about the weather, all the 'political' features, and `unacast_session_count` 

In [1]:
#Download libraries
import pandas as pd
import altair as alt
import numpy as np
import matplotlib.pyplot as plt

In [2]:
#Download the data
df = pd.read_csv('../data/old_train_data.zip')
df.head()

external_id  month  year  monthly_number_of_sessions  \
0     1804425      8  2018                           0   
1     1812706      2  2019                           0   
2     1812706      3  2019                           0   
3     1812706     11  2018                           0   
4     1812706      9  2018                           0   

   monthly_unique_sessions  monthly_repeated_sessions  \
0                        0                          0   
1                        0                          0   
2                        0                          0   
3                        0                          0   
4                        0                          0   

   monthly_avg_length_of_session  monthly_avg_light_activity  \
0                            0.0                         0.0   
1                            0.0                         0.0   
2                            0.0                         0.0   
3                            0.0                         0.0   
4                            0.0                         0.0   

   monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
0                            0.0                            0.0  ...   
1                            0.0                            0.0  ...   
2                            0.0                            0.0  ...   
3                            0.0                            0.0  ...   
4                            0.0                            0.0  ...   

   avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
0            0.0             0.0             0.0                0.0   
1            0.0             0.0             0.0                0.0   
2            0.0             0.0             0.0                0.0   
3            0.0             0.0             0.0                0.0   
4            0.0             0.0             0.0                0.0   

   perfect_days  unacast_session_count     hpi  \
0           0.0                   90.0  244.20   
1           4.0                   27.0  258.95   
2           4.0                   27.0  258.95   
3           3.0                   24.0  258.95   
4           0.0                   12.0  258.95   

   state_and_local_amount_per_capita  state_amount_per_capita  \
0                           0.157475                 0.009783   
1                           0.157475                 0.009783   
2                           0.157475                 0.009783   
3                           0.157475                 0.009783   
4                           0.157475                 0.009783   

   local_amount_per_capita  
0                 0.147692  
1                 0.147692  
2                 0.147692  
3                 0.147692  
4                 0.147692  

[5 rows x 861 columns]

## Study of the NaNs is the temperature data

In [242]:
df[df['avg_wind_12_above'].isnull()]

external_id  month  year  monthly_number_of_sessions  \
556     1901754     11  2018                           0   

     monthly_unique_sessions  monthly_repeated_sessions  \
556                        0                          0   

     monthly_avg_length_of_session  monthly_avg_light_activity  \
556                            0.0                         0.0   

     monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
556                            0.0                            0.0  ...   

     avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
556            NaN             NaN             NaN                NaN   

     perfect_days  unacast_session_count     hpi  \
556           NaN                    8.0  470.64   

     state_and_local_amount_per_capita  state_amount_per_capita  \
556                           0.173926                 0.019939   

     local_amount_per_capita  
556                 0.153987  

[1 rows x 861 columns]

There is only one observation with missing values in the temperature data. Let's study the playground that is the closest to this playground (regarding to the latitude and the longitude), and see if they have similar values for the other months.

In [114]:
my_dist = df.iloc[556][['longitude', 'latitude']]
my_dist[0]

-156.388651

In [116]:
dist = np.square(df['longitude'] - my_dist[0] ) +  np.square(df['latitude'] - my_dist[1])

In [125]:
dist.sort_values().head(21)

20458     0.000000
35507     0.000000
35495     0.000000
35477     0.000000
35395     0.000000
30349     0.000000
35390     0.000000
45631     0.000000
10481     0.000000
5621      0.000000
30400     0.000000
5587      0.000000
20510     0.000000
25446     0.000000
503       0.000000
5537      0.000000
40548     0.000000
546       0.000000
25433     0.000000
556       0.000000
4205     62.768606
dtype: float64

Check that the playgrounds have very close longitude and latitude.

In [127]:
df.iloc[4205][['longitude', 'latitude']]

longitude    -149.87
latitude     61.1987
Name: 4205, dtype: object

In [128]:
my_dist

longitude   -156.389
latitude     65.7019
Name: 556, dtype: object

The two playgrounds seem to be very close to each other.

In [9]:
temp = df[df['external_id'].isin(['MR00111851', '1901754'])].sort_values(['year', 'month'])

In [10]:
temp.loc[:,'weather_clear':'avg_wind_12_above']

weather_clear  weather_rain  weather_fog  weather_snow  \
14621           10.0           0.0          2.0          18.0   
41218           18.0           0.0          0.0           8.0   
39675            5.0           2.0          1.0          26.0   
46288           21.0           0.0          0.0           6.0   
6120            28.0           0.0          0.0           0.0   
49835           22.0           7.0          1.0           6.0   
26012           24.0           0.0          0.0           0.0   
34470           10.0          13.0          2.0           3.0   
1153            25.0           0.0          0.0           0.0   
4601            18.0           6.0          1.0           0.0   
11066           25.0           0.0          0.0           0.0   
39674           19.0           9.0          2.0           0.0   
1154            12.0           0.0          0.0           0.0   
29553            8.0          13.0          0.0           0.0   
36236           27.0           0.0          0.0           0.0   
49834           10.0           8.0          5.0           0.0   
4604            12.0           8.0          1.0           5.0   
30966           16.0           0.0          0.0           1.0   
34469            NaN           NaN          NaN           NaN   
36235           21.0           0.0          0.0           3.0   
9648             7.0           0.0          8.0          24.0   
11065            9.0           0.0          0.0          15.0   
6119            25.0           0.0          0.0           2.0   
49831           10.0           1.0          7.0          19.0   
6118            20.0           0.0          0.0           7.0   
9645             6.0           3.0          6.0          21.0   
9647            12.0           4.0          5.0          18.0   
41219           29.0           0.0          0.0           1.0   
4603            16.0           7.0          3.0          12.0   
26011           21.0           0.0          0.0           6.0   
16096           25.0           0.0          0.0           0.0   
49832           12.0          18.0          1.0           1.0   
1152            30.0           0.0          0.0           0.0   
19497           22.0           8.0          1.0           0.0   
4602            19.0          11.0          2.0           0.0   
26010           27.0           0.0          0.0           0.0   
20989           30.0           0.0          0.0           0.0   
49833           18.0          11.0          2.0           0.0   
9646             0.0           4.0          0.0           0.0   
20988           17.0           0.0          0.0           0.0   

       weather_thunder  weather_cloudy  weather_other  temp_avg_35_below  \
14621              0.0            16.0            0.0               28.0   
41218              0.0             7.0            0.0               28.0   
39675              0.0            25.0            0.0               31.0   
46288              0.0             8.0            0.0               29.0   
6120               0.0             2.0            0.0                5.0   
49835              0.0             7.0            0.0               24.0   
26012              0.0             7.0            0.0                0.0   
34470              0.0            13.0            0.0                3.0   
1153               0.0             5.0            0.0                0.0   
4601               0.0             6.0            0.0                0.0   
11066              0.0             8.0            0.0                0.0   
39674              0.0             8.0            0.0                0.0   
1154               0.0            20.0            0.0                0.0   
29553              0.0            13.0            0.0                0.0   
36236              0.0             4.0            0.0                0.0   
49834              0.0             4.0            0.0                0.0   
4604               0.0             9.0           

The values don't seem to be similar between the two playgrounds. I would just impute the missing values with the means of the values for the same playground, for the previous and the following month.

## Find states with missing values for the 'political' data

In [152]:
df[df['Democrats_12_Votes'].isnull()]['state'].unique()

array(['Alaska'], dtype=object)

In [154]:
df[df['Republicans_2016'].isnull()]['state'].unique()

array(['Alaska'], dtype=object)

In [158]:
df[df['Green_2016'].isnull()]['state'].unique()

array(['Georgia', 'North Carolina', 'Nevada', 'Indiana', 'South Dakota',
       'Alaska', 'Oklahoma'], dtype=object)

## Study of the `unacast_session_count` 

Let's study the playgrounds that only have missing values until a given month.

In [5]:
#Let's create an array that contains the months and the years we are studying
test_df = df[df['external_id'] == 'MR00117367'].loc[:, 'month':'year']
test_df = test_df.sort_values(by=['year', 'month']).reset_index()
time = test_df[['month', 'year']].to_numpy()
time

array([[   2, 2018],
       [   3, 2018],
       [   4, 2018],
       [   5, 2018],
       [   6, 2018],
       [   7, 2018],
       [   8, 2018],
       [   9, 2018],
       [  10, 2018],
       [  11, 2018],
       [  12, 2018],
       [   1, 2019],
       [   2, 2019],
       [   3, 2019],
       [   4, 2019],
       [   5, 2019],
       [   6, 2019],
       [   7, 2019],
       [   8, 2019],
       [   9, 2019]])

In [30]:
#my_dict will contains the id of the playgrounds with missing values until a given month and a given year.
#Initialization of the dictionnary with the values from February 2018 
my_dict = {'month':[time[0][0]], 'year':[time[0][1]], 'playgrounds':[list(df[df['unacast_session_count'].isnull()].query('year==2018 & month==2')['external_id'].values)]}
for i in range(len(time)-1):
    list_play = list(df[df['unacast_session_count'].isnull()].query('year==' + str(time[i+1][1]) + ' & month==' + str(time[i+1][0]))['external_id'].values)
    #Check that the id playground is in the list of playgrounds with missing target for the month under study, and all the privious months
    name_play = list(set(list_play) & set(my_dict['playgrounds'][-1]))

    my_dict['month'].append(time[i+1][0])
    my_dict['year'].append(time[i+1][1])
    my_dict['playgrounds'].append(name_play)


In [31]:
playgrounds = pd.DataFrame(my_dict)

#Add the count of playgrounds that only have missing values until the given month
playgrounds['count'] = list(map(lambda x : len(x), playgrounds['playgrounds']))

In [32]:
playgrounds

month  year                                        playgrounds  count
0       2  2018  [1811740, 5ba4de98-f4cc-4c60-9439-262b3405d628...    124
1       3  2018  [MR00106561, FM00170998, FM00171867, 1809195, ...     43
2       4  2018  [MR00101262, MR00105990, FM00171867, FM0017099...     17
3       5  2018  [MR00101262, MR00105990, FM00171867, FM0017099...      7
4       6  2018   [MR00101262, MR00095127, FM00171867, MR00105990]      4
5       7  2018   [MR00101262, MR00095127, FM00171867, MR00105990]      4
6       8  2018   [MR00101262, MR00095127, FM00171867, MR00105990]      4
7       9  2018   [MR00101262, MR00095127, FM00171867, MR00105990]      4
8      10  2018   [MR00101262, MR00095127, FM00171867, MR00105990]      4
9      11  2018               [MR00101262, MR00095127, MR00105990]      3
10     12  2018                           [MR00101262, MR00105990]      2
11      1  2019                                       [MR00101262]      1
12      2  2019                                       [MR00101262]      1
13      3  2019                                       [MR00101262]      1
14      4  2019                                       [MR00101262]      1
15      5  2019                                                 []      0
16      6  2019                                                 []      0
17      7  2019                                                 []      0
18      8  2019                                                 []      0
19      9  2019                                                 []      0

The dataframe above has to be understood in the following way : if we take for example the row with the index 7, it means that the playgrounds whose external_id belongs in [MR00101262, FM00171867, MR00105990, MR00095127] only have missing target value until September 2018 (so for those playgrounds, the `unicast_session_count` is NaN for all the months between February 2018 and September 2018).

In [220]:
playgrounds['count'].sum()

220

Let's delete the corresponding rows from the original dataset (as we may suspect that those values are NaNs because the playground didn't exist yet).

In [33]:
my_df = df.copy()
for i in range(playgrounds.shape[0]):
    year = playgrounds['year'][i]
    month = playgrounds['month'][i]
    ids = playgrounds['playgrounds'][i]
    my_df = my_df[((my_df.year != year) | (my_df.month != month) | (my_df.external_id.isin(ids) == False))]

In [34]:
my_df

external_id  month  year  monthly_number_of_sessions  \
0         1804425      8  2018                           0   
1         1812706      2  2019                           0   
2         1812706      3  2019                           0   
3         1812706     11  2018                           0   
4         1812706      9  2018                           0   
...           ...    ...   ...                         ...   
50115  MR00109612      2  2019                           8   
50116  MR00109612      9  2018                           0   
50117  MR00109612      4  2018                           0   
50118  MR00108096      7  2019                           0   
50119  MR00108096      5  2019                           0   

       monthly_unique_sessions  monthly_repeated_sessions  \
0                            0                          0   
1                            0                          0   
2                            0                          0   
3                            0                          0   
4                            0                          0   
...                        ...                        ...   
50115                        1                          7   
50116                        0                          0   
50117                        0                          0   
50118                        0                          0   
50119                        0                          0   

       monthly_avg_length_of_session  monthly_avg_light_activity  \
0                                0.0                         0.0   
1                                0.0                         0.0   
2                                0.0                         0.0   
3                                0.0                         0.0   
4                                0.0                         0.0   
...                              ...                         ...   
50115                       773000.0                         0.0   
50116                            0.0                         0.0   
50117                            0.0                         0.0   
50118                            0.0                         0.0   
50119                            0.0                         0.0   

       monthly_avg_moderate_activity  monthly_avg_vigorous_activity  ...  \
0                                0.0                            0.0  ...   
1                                0.0                            0.0  ...   
2                                0.0                            0.0  ...   
3                                0.0                            0.0  ...   
4                                0.0                            0.0  ...   
...                              ...                            ...  ...   
50115                       113000.0                       171000.0  ...   
50116                            0.0                            0.0  ...   
50117                            0.0                            0.0  ...   
50118                            0.0                            0.0  ...   
50119                            0.0                            0.0  ...   

       avg_wind_9_10  avg_wind_10_11  avg_wind_11_12  avg_wind_12_above  \
0                0.0             0.0             0.0                0.0   
1                0.0             0.0             0.0                0.0   
2                0.0             0.0             0.0                0.0   
3                0.0             0.0             0.0                0.0   
4                0.0             0.0             0.0                0.0   
...              ...             ...             ...                ...   
50115            0.0             0.0             0.0                0.0   
50116            0.0             0.0             0.0                0.0   
50117            0.0             0.0             0.0                0.0   
50118            0.0             0.0            

In [243]:
#Check if the dimensions of my_df is right
df.shape[0] - playgrounds['count'].sum() == my_df.shape[0]

True

In [238]:
my_df[my_df['unacast_session_count'].isnull()]['external_id'].value_counts()

e5bdd7bc-325a-4f88-ad22-a3a820118b68    15
MR00110941                              15
MR00117367                              13
1805183                                  7
MR00118229                               7
                                        ..
MR00108479                               1
1810710                                  1
1810866                                  1
FM00168120                               1
1806104                                  1
Name: external_id, Length: 164, dtype: int64

In [44]:
df[df['unacast_session_count'].isnull()]['external_id'].value_counts()

MR00101262                              19
MR00105990                              18
MR00110941                              18
MR00117367                              16
e5bdd7bc-325a-4f88-ad22-a3a820118b68    15
                                        ..
FM00170586                               1
1805327                                  1
MR00106063                               1
1806831                                  1
FM00171112                               1
Name: external_id, Length: 222, dtype: int64

We can see that once we delete the rows that correspond to non-existing playgrounds, we only have 164 playgrounds with NaNs values for the `unacast_session_count`.

Now, let's look at some particular playgrounds to see if we can notice if we can understand where the NaNs values come from.

In [47]:
my_df[my_df['external_id'] == '1805183'][['month',  'year', 'unacast_session_count']].sort_values(['year', 'month'])

month  year  unacast_session_count
7089       6  2018                    1.0
11995      7  2018                    NaN
2068       8  2018                    1.0
11996      9  2018                    NaN
37167     10  2018                    NaN
26971     11  2018                    1.0
7092      12  2018                    NaN
7090       1  2019                    NaN
11994      2  2019                    NaN
7091       3  2019                    NaN
2067       4  2019                    3.0
47244      5  2019                    8.0
47243      6  2019                   11.0
16974      7  2019                    9.0
16975      8  2019                    1.0
16973      9  2019                    2.0

In [42]:
my_df[my_df['external_id'] == 'MR00108479'][['month',  'year', 'unacast_session_count']].sort_values(['year', 'month'])

month  year  unacast_session_count
1333       2  2018                   54.0
16309      3  2018                   49.0
26232      4  2018                   32.0
6344       5  2018                   27.0
41419      6  2018                   31.0
6343       7  2018                   27.0
26233      8  2018                   18.0
36459      9  2018                   25.0
31181     10  2018                   35.0
6342      11  2018                   60.0
1334      12  2018                   27.0
31180      1  2019                   57.0
26231      2  2019                   45.0
11284      3  2019                   48.0
41420      4  2019                   28.0
11283      5  2019                   51.0
41418      6  2019                   28.0
36458      7  2019                    NaN
1335       8  2019                  118.0
46474      9  2019                   47.0

In [51]:
my_df[(my_df['month'] == 2)&(my_df['year']==2019) & (my_df['unacast_session_count'].isnull())]

external_id  month  year  \
1064                             MR00110011      2  2019   
3879   3bcfc097-2470-4a67-82ec-b00dcc0cb204      2  2019   
6752                                1902864      2  2019   
7899                             FM00175134      2  2019   
10754                               1807213      2  2019   
10851                            MR00119603      2  2019   
11994                               1805183      2  2019   
12144  342b9a09-4499-409a-978a-1525ee0f8ed3      2  2019   
13122                            MR00116613      2  2019   
13398                            MR00107074      2  2019   
14606                               1809130      2  2019   
14744                               1809087      2  2019   
15678                            FM00164842      2  2019   
16129                            MR00110377      2  2019   
17017                               1902483      2  2019   
17575                            FM00167820      2  2019   
17599                            MR00109740      2  2019   
17687                            MR00109724      2  2019   
17718                            MR00098684      2  2019   
18288                               1803563      2  2019   
19240                            FM00168198      2  2019   
20122                            FM00176895      2  2019   
21711                            MR00118152      2  2019   
22875                            MR00105748      2  2019   
23847                            MR00107328      2  2019   
28043                            FM00162852      2  2019   
28186                            MR00105777      2  2019   
29320                            MR00116453      2  2019   
30014                            MR00115625      2  2019   
30189                            MR00110941      2  2019   
31219                            MR00104270      2  2019   
31417                            MR00112155      2  2019   
33528  e5bdd7bc-325a-4f88-ad22-a3a820118b68      2  2019   
35096                            FM00173873      2  2019   
35450  5ba4de98-f4cc-4c60-9439-262b3405d628      2  2019   
35548                            FM00166894      2  2019   
35658                            MR00107303      2  2019   
38186                            MR00116410      2  2019   
38225                            MR00098899      2  2019   
41782                               1900839      2  2019   
42020                            MR00108289      2  2019   
42202                            FM00163789      2  2019   
42760                            FM00168960      2  2019   
43247                            MR00117367      2  2019   
44086                            MR00098614      2  2019   
44161                            MR00105746      2  2019   
45958                            MR00119521      2  2019   
46530                            FM00164052      2  2019   
46533                            MR00108384      2  2019   
48805                            MR00098983      2  2019   
49281                            FM00166318      2  2019   
49840                            MR00105990      2  2019   
49860                               1809358      2  2019   

       monthly_number_of_sessions  monthly_unique_sessions  \
1064                            0                        0   
3879                            0                        0   
6752                            0                        0   
7899                            0                        0   
10754                           0                        0   
10851                           0                        0   
11994                           0                        0   
12144                           0                        0   
13122                           0                        0   
13398                           0                        0   
14606                           0                        0   
14744                           0              

A lot of missing values seem to happen during the winter, or in playgrounds that have only a very low number of unacast sessions. This may suggests that a lot of NaNs actually correspond to 0s. However, there are also some playgrounds with missing values that don't correspond to the winter, and where there are a lot of visits : those ones may just correspond to an anomalie, and it's harder to find a value we could use to impute them.

## Conclusion

- For the 'weather data', only one row have missing values, thus I would recommand to impute the missing values with the means of the values for the same playground, for the previous and the following month. Looking at the 'closest' playground (from a distance point of view) doesn't seem to be a good idea.

- For the 'political data', only a few states misses values, so it wouldn't be hard to find those values online.

- For the `unacast_session_count`, we found that a third of the missing values happen at the beggining of a playground's timeline. Therefore, we thought that we could delete those rows as they may correspond to a time where the unacast sessions were not collected for the given playground. Regarding the remaining NaNs values, most of them seem to correspond to 0s, while a few of them seem to be anomalies, which are going to be a lot harder to impute.